### ISSUE TITLE
Code Quality: Dispose COM pointers automatically using IDisposable

### ISSUE THREAD

**Issue by 0x5bfa** [files-community/Files#16071]
### Description

I sometimes unbelievably forget to dispose unmanaged resources. If you do that for COM interfaces, it might cause a severe issue leaking memory.

Struct impl:
```c#
public unsafe struct ComPtr<T> : IDisposable
    where T : unmanaged, IUnknown
{
    private _ptr;

    public ComPtr(T* ptr)
    {
        _ptr = ptr;
    }

    public Dispose()
    {
        _ptr->Release();
    }
}
```

Usage:
```c#
public void Do()
{
    try
    {
        using ComPtr<IFileOperation> pFileOperation = default;

        HRESULT hRes = CoCreateInstance(..., *(void**)pFileOperation.GetAddressOf);
    }
    catch (Exception ex)
    {
    }
}
``` 

### Concerned code

All com interface pointers.

### Gains

- prevent memory leak

### Requirements

- Implement a struct that implements IDisposable and use it through using statement

**Comment by 0x5bfa** [comment]
This was implemented in #15911 and it works well.
I'm considering cherry picking into another PR to merge it first.

**Comment by 0x5bfa** [comment]
I'll consider to use [IDynamicInterfaceCastable](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.idynamicinterfacecastable?view=net-8.0) that would supersede ComQIPtr after .NET 9 comes out.

In the meantime, I'd propose:

```c#
readonly ComPtr<U> QueryFor<U>() where U : unmanaged, IComIID
{
    ComPtr<U> ptr = default;
    _ptr->QueryInterface(&U.Guid, (void**)ptr.GetAddressOf());
    return ptr;
}

// or

readonly HRESULT QueryFor<U>(U** ptr) where U : unmanaged, IComIID
{
    _ptr->QueryInterface(&U.Guid, (void**)ptr);
}
```

```c#
using ComPtr<IShellItem> pShellItem = default;
using ComPtr<IShellItem2> pShellItem2 = pShellItem.QueryFor<IShellItem2>();

// or

using ComPtr<IShellItem> pShellItem = default;
using ComPtr<IShellItem2> pShellItem2 = default;
pShellItem.QueryFor<IShellItem2>(pShellItem2.GetAddressOf());
```

Not sure which is better for now, but I'll implement either of them.
